src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -325,13 +329,56 @@ internal override void WriteWithQuotes(Utf8JsonWriter writer, T value, JsonSeria
 
             ulong key = ConvertToUInt64(value);
 
-            if (_nameCache.TryGetValue(key, out JsonEncodedText formatted))
+            JsonEncodedText formatted;
+            string original;
+
+            if (options.DictionaryKeyPolicy != null && !state.Current.IgnoreDictionaryKeyPolicy)
+            {
+                if (_dictionaryKeyPolicyCache.Value.TryGetValue(key, out formatted))
```

Comment:
I think this should be:

```suggestion
            if (options.DictionaryKeyPolicy != null)
            {
                Debug.Assert(!state.Current.IgnoreDictionaryKeyPolicy);
            
                if (_dictionaryKeyPolicyCache.Value.TryGetValue(key, out formatted))
```

I believe that boolean is only set when processing extension data dictionary keys which are string-only (handled by a different converter).